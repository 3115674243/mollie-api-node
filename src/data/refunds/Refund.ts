import { RefundData } from './data';
import refundHelpers from './helpers';
import Seal from '../../types/Seal';
import OrderLine, { injectPrototypes as injectOrderLinePrototypes } from '../orders/orderlines/OrderLine';

type Refund = Seal<RefundData & { lines?: OrderLine[] }, typeof refundHelpers>;

export default Refund;

export function injectPrototypes(input: RefundData): Refund {
  let lines: Refund['lines'];
  if (input.lines != undefined) {
    lines = input.lines.map(injectOrderLinePrototypes);
  }
  return Object.assign(Object.create(refundHelpers), input, {
    lines,
  });
}
