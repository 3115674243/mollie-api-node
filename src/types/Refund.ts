declare namespace Mollie {
  type IRefund = IPaymentRefund | IOrderRefund;

  namespace Refund {
    enum Status {
      queued = 'queued',
      pending = 'pending',
      processing = 'processing',
      refunded = 'refunded',
      failed = 'failed',
    }

    namespace Params {
      interface IList {}
    }

    namespace Callback {
      type List = (error: any, refunds: Mollie.List<IRefund>) => void;
    }
  }
}
