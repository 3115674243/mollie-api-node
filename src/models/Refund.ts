import { isPlainObject } from 'lodash';

import Model from '../model';
import { RefundStatus } from '../types/refund';
import { IRefund } from '../types/payment/refund';
import Payment from './Payment';

/**
 * The `Refund` model
 *
 * {@link IRefund}
 */
export default class Refund extends Model implements IRefund {
  public static resourcePrefix = 're_';

  public resource = 'refund';
  public id = null;
  public amount = {
    currency: null,
    value: null,
  };
  public settlementAmount = null;
  public description = null;
  public status = null;
  public createdAt = null;
  public paymentId = null;
  public _links = {
    payment: null,
    settlement: null,
    order: null,
    self: null,
    documentation: null,
  };
  public lines = null;
  public _embedded = null;
  public orderId = null;

  /**
   * Refund constructor
   *
   * @public ✓ This constructor is part of the public API
   */
  public constructor(props?: Partial<IRefund>) {
    super();

    Object.assign(this, props);

    if (this._embedded != null && isPlainObject(this._embedded)) {
      if (Array.isArray(this._embedded.payments)) {
        // eslint-disable-next-line array-callback-return
        this._embedded.payments.map((payment, key, payments) => {
          // eslint-disable-next-line no-param-reassign
          payments[key] = new Payment(payment);
        });
      }
    }
  }

  /**
   * The refund is queued until there is enough balance to process te refund.
   * You can still cancel the refund.
   *
   * @returns {boolean}
   *
   * @public ✓ This method is part of the public API
   */
  public isQueued(): boolean {
    return this.status === RefundStatus.queued;
  }

  /**
   * The refund will be sent to the bank on the next business day. You can still cancel the refund.
   *
   * @public ✓ This method is part of the public API
   */
  public isPending(): boolean {
    return this.status === RefundStatus.pending;
  }

  /**
   * The refund has been sent to the bank. The refund amount will be transferred to the consumer
   * account as soon as possible.
   *
   * @public ✓ This method is part of the public API
   */
  public isProcessing(): boolean {
    return this.status === RefundStatus.processing;
  }

  /**
   * The refund amount has been transferred to the consumer.
   *
   * @public ✓ This method is part of the public API
   */
  public isRefunded(): boolean {
    return this.status === RefundStatus.refunded;
  }

  /**
   * The refund has failed during processing.
   *
   * @public ✓ This method is part of the public API
   */
  public isFailed(): boolean {
    return this.status === RefundStatus.failed;
  }
}
