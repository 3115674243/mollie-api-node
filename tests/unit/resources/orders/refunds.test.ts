import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import OrdersRefunds from '../../../../src/resources/orders/refunds';
import Refund from '../../../../src/models/refund';

import response from '../../__stubs__/orders_refunds.json';

const mock = new MockAdapter(axios);

const props = {
  id: 're_4qqhO89gsT',
  orderId: 'ord_stTC2WHAuS',
  lines: [
    {
      id: 'odl_dgtxyl',
      quantity: 1
    }
  ],
};

describe('orders_refunds', () => {
  let ordersRefunds;
  beforeEach(() => {
    ordersRefunds = new OrdersRefunds(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(OrdersRefunds.resource).toBe('orders_refunds');
    expect(OrdersRefunds.model).toBe(Refund);
  });

  describe('.create()', () => {
    mock.onPost(`/orders/${props.orderId}/refunds`).reply(200, response._embedded.refunds[0]);

    it('should return a refund instance', () =>
      ordersRefunds.create(props).then(result => {
        expect(result).toBeInstanceOf(Refund);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      ordersRefunds.create(props, (err, result) => {
        expect(result).toBeInstanceOf(Refund);
        expect(result).toMatchSnapshot();
        done();
      });
    });
  });

  describe('.all()', () => {
    mock.onGet(`/orders/${props.orderId}/refunds`).reply(200, response);

    it('should return a list of all order refunds', () =>
      ordersRefunds.all({ orderId: props.orderId }).then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should throw an error if "orderId" is not set', () => {
      const getRefunds = () => ordersRefunds.all();

      expect(getRefunds).toThrowError(TypeError);
    });

    it('should work with a callback', done => {
      ordersRefunds
        .withParent({
          resource: 'order',
          id: props.orderId,
        })
        .all((err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        });
    });
  });
});
