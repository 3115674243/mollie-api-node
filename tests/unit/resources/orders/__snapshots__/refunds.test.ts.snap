// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orders_refunds .all() should return a list of all order refunds 1`] = `
Array [
  Refund {
    "_embedded": null,
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/refunds-api/get-refund",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "payment": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg/refunds/re_4qqhO89gsT",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "698.00",
    },
    "createdAt": "2018-03-14T17:09:02.0Z",
    "description": "Required quantity not in stock, refunding one photo book.",
    "id": "re_4qqhO89gsT",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_stTC2WHAuS",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "paid",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "orderId": "ord_stTC2WHAuS",
    "paymentId": "tr_WDqYK6vllg",
    "resource": "refund",
    "settlementAmount": null,
    "status": "processing",
  },
]
`;

exports[`orders_refunds .all() should work with a Promise and with .withParent() 1`] = `
Array [
  Refund {
    "_embedded": null,
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/refunds-api/get-refund",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "payment": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg/refunds/re_4qqhO89gsT",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "698.00",
    },
    "createdAt": "2018-03-14T17:09:02.0Z",
    "description": "Required quantity not in stock, refunding one photo book.",
    "id": "re_4qqhO89gsT",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_stTC2WHAuS",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "paid",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "orderId": "ord_stTC2WHAuS",
    "paymentId": "tr_WDqYK6vllg",
    "resource": "refund",
    "settlementAmount": null,
    "status": "processing",
  },
]
`;

exports[`orders_refunds .all() should work with a callback 1`] = `
Array [
  Refund {
    "_embedded": null,
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/refunds-api/get-refund",
        "type": "text/html",
      },
      "order": Object {
        "href": "https://api.mollie.com/v2/orders/ord_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "payment": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg/refunds/re_4qqhO89gsT",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "698.00",
    },
    "createdAt": "2018-03-14T17:09:02.0Z",
    "description": "Required quantity not in stock, refunding one photo book.",
    "id": "re_4qqhO89gsT",
    "lines": Array [
      Object {
        "_links": Object {
          "imageUrl": Object {
            "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
            "type": "text/html",
          },
          "productUrl": Object {
            "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
            "type": "text/html",
          },
        },
        "createdAt": "2018-08-02T09:29:56+00:00",
        "discountAmount": Object {
          "currency": "EUR",
          "value": "100.00",
        },
        "id": "odl_dgtxyl",
        "name": "LEGO 42083 Bugatti Chiron",
        "orderId": "ord_stTC2WHAuS",
        "quantity": 1,
        "resource": "orderline",
        "sku": "5702016116977",
        "status": "paid",
        "totalAmount": Object {
          "currency": "EUR",
          "value": "299.00",
        },
        "type": "physical",
        "unitPrice": Object {
          "currency": "EUR",
          "value": "399.00",
        },
        "vatAmount": Object {
          "currency": "EUR",
          "value": "51.89",
        },
        "vatRate": "21.00",
      },
    ],
    "orderId": "ord_stTC2WHAuS",
    "paymentId": "tr_WDqYK6vllg",
    "resource": "refund",
    "settlementAmount": null,
    "status": "processing",
  },
]
`;

exports[`orders_refunds .create() should return a refund instance 1`] = `
Refund {
  "_embedded": null,
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/refunds-api/get-refund",
      "type": "text/html",
    },
    "order": Object {
      "href": "https://api.mollie.com/v2/orders/ord_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "payment": Object {
      "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg/refunds/re_4qqhO89gsT",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "698.00",
  },
  "createdAt": "2018-03-14T17:09:02.0Z",
  "description": "Required quantity not in stock, refunding one photo book.",
  "id": "re_4qqhO89gsT",
  "lines": Array [
    Object {
      "_links": Object {
        "imageUrl": Object {
          "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
          "type": "text/html",
        },
        "productUrl": Object {
          "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
          "type": "text/html",
        },
      },
      "createdAt": "2018-08-02T09:29:56+00:00",
      "discountAmount": Object {
        "currency": "EUR",
        "value": "100.00",
      },
      "id": "odl_dgtxyl",
      "name": "LEGO 42083 Bugatti Chiron",
      "orderId": "ord_stTC2WHAuS",
      "quantity": 1,
      "resource": "orderline",
      "sku": "5702016116977",
      "status": "paid",
      "totalAmount": Object {
        "currency": "EUR",
        "value": "299.00",
      },
      "type": "physical",
      "unitPrice": Object {
        "currency": "EUR",
        "value": "399.00",
      },
      "vatAmount": Object {
        "currency": "EUR",
        "value": "51.89",
      },
      "vatRate": "21.00",
    },
  ],
  "orderId": "ord_stTC2WHAuS",
  "paymentId": "tr_WDqYK6vllg",
  "resource": "refund",
  "settlementAmount": null,
  "status": "processing",
}
`;

exports[`orders_refunds .create() should work with a callback 1`] = `
Refund {
  "_embedded": null,
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/refunds-api/get-refund",
      "type": "text/html",
    },
    "order": Object {
      "href": "https://api.mollie.com/v2/orders/ord_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "payment": Object {
      "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_WDqYK6vllg/refunds/re_4qqhO89gsT",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "698.00",
  },
  "createdAt": "2018-03-14T17:09:02.0Z",
  "description": "Required quantity not in stock, refunding one photo book.",
  "id": "re_4qqhO89gsT",
  "lines": Array [
    Object {
      "_links": Object {
        "imageUrl": Object {
          "href": "https://sh-s7-live-s.legocdn.com/is/image//LEGO/42083_alt1?$main$",
          "type": "text/html",
        },
        "productUrl": Object {
          "href": "https://shop.lego.com/nl-NL/Bugatti-Chiron-42083",
          "type": "text/html",
        },
      },
      "createdAt": "2018-08-02T09:29:56+00:00",
      "discountAmount": Object {
        "currency": "EUR",
        "value": "100.00",
      },
      "id": "odl_dgtxyl",
      "name": "LEGO 42083 Bugatti Chiron",
      "orderId": "ord_stTC2WHAuS",
      "quantity": 1,
      "resource": "orderline",
      "sku": "5702016116977",
      "status": "paid",
      "totalAmount": Object {
        "currency": "EUR",
        "value": "299.00",
      },
      "type": "physical",
      "unitPrice": Object {
        "currency": "EUR",
        "value": "399.00",
      },
      "vatAmount": Object {
        "currency": "EUR",
        "value": "51.89",
      },
      "vatRate": "21.00",
    },
  ],
  "orderId": "ord_stTC2WHAuS",
  "paymentId": "tr_WDqYK6vllg",
  "resource": "refund",
  "settlementAmount": null,
  "status": "processing",
}
`;
